### PR TITLE
remove mars.d use of lib

### PR DESCRIPTION
**Pull request by thewilsonator**
dmd.lib is now solely imported by glue.d

### CODE REVIEW

File: src/dmd/glue.d [hunk 1]
```diff
@@ -72,12 +72,71 @@ extern (C++):
 alias symbols = Array!(Symbol*);
 alias toSymbol = dmd.tocsym.toSymbol;
 
+void generateCodeAndWrite(ref Modules modules, ref Strings libmodules, ref Params params)
```
Comment (thewilsonator) [inline src/dmd/glue.d:75]: names are hard, better ideas appreciated.

File: src/dmd/mars.d [hunk 2]
```diff
@@ -24,7 +24,7 @@ import core.stdc.string;
 
 import dmd.arraytypes;
 import dmd.astcodegen;
-import dmd.gluelayer;
+import dmd.glue : generateCodeAndWrite;
```
Comment (ibuclaw) [inline src/dmd/mars.d:27]: Judging by the failures. Put this in version(!NoMain)

Comment (thewilsonator) [reply src/dmd/mars.d]: Thanks, Will try

**Comment by ibuclaw** [comment]
Progress. 
```
Error: program killed by signal 11
```

**Comment by thewilsonator** [comment]
Hmm, that fixed some things. `dmd-unittest` still segfaults.

File: src/dmd/glue.d [hunk 3]
```diff
@@ -67,17 +68,87 @@ import dmd.toobj;
...
+version (NoMain)
+{
+    void generateCodeAndWrite(Module[] modules, const(char)*[] libmodules,
+                              const(char)[] libname, const(char)[] objdir,
+                              bool lib, bool obj, bool oneobj, bool multiobj,
+                              bool verbose) {}
+}
+else
+{
+void generateCodeAndWrite(Module[] modules, const(char)*[] libmodules,
```
Comment (ibuclaw) [inline src/dmd/glue.d:82]:
```suggestion
void generateCodeAndWrite(Module[] modules, const(char)*[] libmodules,
```
The glue module isn't astbase-friendly, so this shouldn't be needed I guess.

File: src/dmd/mars.d [hunk 4]
```diff
@@ -464,8 +463,6 @@ private int tryMain(size_t argc, const(char)** argv, ref Param params)
     if (global.errors)
         removeHdrFilesAndFail(params, modules);
 
-    backend_init();
```
Comment (ibuclaw) [inline src/dmd/mars.d:467]: I assumed that semantic depended on this somewhere. Maybe typinf or any other of the gluey frontend modules.

Comment (thewilsonator) [reply src/dmd/mars.d:467]: Hmm, yes. It shouldn't though

Comment (thewilsonator) [reply src/dmd/mars.d:467]: So it seems that at least semantic1 requires the backend to be initialised: https://github.com/dlang/dmd/pull/12857 any ideas?

Comment (thewilsonator) [reply src/dmd/mars.d:467]: I could always hoist `backend_init` back out of the new code generation function, but there is something very wrong about semantic analysis requiring an initialised backend.

Comment (ibuclaw) [reply src/dmd/mars.d:467]: @thewilsonator I changed it to a pointer in [this test](https://github.com/ibuclaw/dmd/commit/c30e57b2637d8fb77703ad00ba8d4e55232f1e4c), and compiler segfaults in `dmd.iasmdmd`.

Comment (thewilsonator) [reply src/dmd/mars.d:467]: Neat, what were the first few function in the backtrace?

Comment (thewilsonator) [reply src/dmd/mars.d:467]: Oh, there's only one occurrence nvm

Comment (ibuclaw) [reply src/dmd/mars.d:467]: @thewilsonator there's more than one occurrence.  See the use of `CodeBuilder` in iasmdmd, and all of its methods.

Comment (ibuclaw) [reply src/dmd/mars.d:467]: After swapping `config.target_cpu` for `target.cpu`, backtrace is now:
```
#0  0x0000555555b49717 in simplify_code(code*) (c=0x555558d62d78) at src/dmd/backend/cod3.d:6360                                                [1197/1347]
#1  0x0000555555a94dad in CodeBuilder::gen(code*) (this=..., cs=0x7ffffffef490) at src/dmd/backend/codebuilder.d:125
#2  0x0000555555a952b5 in CodeBuilder::genlinnum(Srcpos) (this=..., srcpos=...) at src/dmd/backend/codebuilder.d:304
#3  0x0000555555a6a466 in dmd.iasmdmd.asm_emit(dmd.globals.Loc, uint, dmd.backend.iasm.PTRNTAB, dmd.backend.iasm.OP*, dmd.iasmdmd.OPND[]).finalizeCode(dmd
#4  0x0000555555a6a074 in dmd.iasmdmd.asm_emit(dmd.globals.Loc, uint, dmd.backend.iasm.PTRNTAB, dmd.backend.iasm.OP*, dmd.iasmdmd.OPND[]) (opnds=..., pop=
#5  0x0000555555a65b13 in dmd.iasmdmd.inlineAsmSemantic(dmd.statement.InlineAsmStatement, dmd.dscope.Scope*) (sc=0x7fffe4f4e060, s=0x7fffe4f676a0) at src/
#6  0x0000555555a65591 in asmSemantic(AsmStatement*, Scope*) (s=0x7fffe4f67040, sc=0x7fffe4f4e060) at src/dmd/iasm.d:47
#7  0x0000555555a215b9 in StatementSemanticVisitor::visit(AsmStatement*) (this=0x7fffffff0148, s=0x7fffe4f67040) at src/dmd/statementsem.d:4265
...
```

Comment (ibuclaw) [reply src/dmd/mars.d:467]: So, IMO, s2ir should be doing the actual code generation.  The semantic pass should just be verifying that all parts of the statement are correct.

Comment (thewilsonator) [reply src/dmd/mars.d:467]: Yes. Will fix that separately.

Comment (thewilsonator) [reply src/dmd/mars.d:467]: Don't let me stop you from doing that though!

Comment (ibuclaw) [reply src/dmd/mars.d:467]: Nah... :-)

**Comment by thewilsonator** [comment]
Restored bracketing `backend_{init, term}`, this is ready for review.

**Comment by thewilsonator** [comment]
ping @ibuclaw

File: src/dmd/glue.d [hunk 5]
```diff
@@ -67,17 +68,77 @@ import dmd.toobj;
 import dmd.typesem;
 import dmd.utils;
 
-extern (C++):
-
 alias symbols = Array!(Symbol*);
 alias toSymbol = dmd.tocsym.toSymbol;
 
+void generateCodeAndWrite(Module[] modules, const(char)*[] libmodules,
```
Comment (dkorpel) [inline src/dmd/glue.d:74]: New `public` symbol needs Ddoc comment

**Review by dkorpel — COMMENTED** [review]
Shorter tryMain, less code, more private, looking good!

File: src/dmd/glue.d (hunk 5 above)
Comment (thewilsonator) [reply src/dmd/glue.d:88]: done

**Review by dkorpel — APPROVED** [review]

**Pull request merged**